src/manifolds/SymplecticStiefel.jl

```diff
@@ -0,0 +1,664 @@
+``ΛΓ^T \in ℝ^{8k \times 8k}`` and ``λγ^T \in ℝ^{4k \times 4k}``.
+"""
+exp(::SymplecticStiefel, p, X)
+
+function exp!(M::SymplecticStiefel{n,k}, q, p, X) where {n,k}
+    Q = SymplecticMatrix(p, X)
+    pT_p = lu(p' * p) # ∈ ℝ^{2k × 2k}
+
+    C = pT_p \ X' # ∈ ℝ^{2k × 2n}
+
+    # Construct A-bar:
+    # First A-term: Q * (p^T * C^T) * Q
+    A_bar = rmul!(lmul!(Q, (p' * C')), Q)
+    A_bar .+= C * p
+
+    # Second A-term, use C-memory:
+    rmul!(C, Q') # C*Q^T -> C
+    C_QT = C
+
+    # Subtract C*Q^T*p*(pT_p)^{-1}*Q:
+    # A_bar is "star-skew symmetric" (A^+ = -A).
+    A_bar .-= rmul!((C_QT * p) / pT_p, Q)
+
+    # Construct H_bar:
+    # First H-term: Q * (C_QT * Q)' * Q -> Q * C' * Q = Q * (X / pT_p) * Q
+    H_bar = rmul!(lmul!(Q, rmul!(C_QT, Q)'), Q)
+    # Subtract second H-term:
+    H_bar .-= p * symplectic_inverse_times(M, p, H_bar)
+
+    # Construct Δ_bar in H_bar-memory:
+    H_bar .+= p * A_bar
+
+    # Rename H_bar -> Δ_bar.
+    Δ_bar = H_bar
+
+    γ_1 = Δ_bar - (1 / 2) .* p * symplectic_inverse_times(M, p, Δ_bar)
+    γ = [γ_1 -p] # ∈ ℝ^{2n × 4k}
+
+    Δ_bar_star = rmul!(Q' * Δ_bar', Q)
+    λ_1 = lmul!(Q', p * Q)
+    λ_2 = (Δ_bar_star .- (1 / 2) .* (Δ_bar_star * p) * λ_1')'
+    λ = [λ_1 λ_2] # ∈ ℝ^{2n × 4k}
+
+    Γ = [λ -γ] # ∈ ℝ^{2n × 8k}
+    Λ = [γ λ] # ∈ ℝ^{2n × 8k}
+
+    # At last compute the (8k × 8k) and (4k × 4k) matrix exponentials:
+    q .= Γ * (exp(Λ' * Γ)[:, (4k + 1):end]) * (exp(λ' * γ)[:, (2k + 1):end])
+    return q
+end
+
+@doc raw"""
+    gradient(::SymplecticStiefel, f, p, backend::RiemannianProjectionBackend)
+    gradient!(::SymplecticStiefel, f, X, p, backend::RiemannianProjectionBackend)
+
+Compute the gradient of
+``f\colon \operatorname{SpSt}(2n, 2k) \rightarrow ℝ``
+at ``p \in \operatorname{SpSt}(2n, 2k)``.
+
+We first compute the embedding gradient ``∇f(p) \in ℝ^{2n \times 2k}`` using
+the [`AbstractRiemannianDiffBackend`](@ref) in the [`RiemannianProjectionBackend`](@ref).
+Then we transform the embedding gradient to the unique tangent vector space element
+``\text{grad}f(p) \in T_p\operatorname{SpSt}(2n, 2k)``
+which fulfills the variational equation
+````math
+    g_p(\text{grad}f(p), X)
+    = \text{D}f(p)
+    = \langle ∇f(p), X \rangle
+    \quad\forall\; X \in T_p\operatorname{SpSt}(2n, 2k).
+````
+The manifold gradient ``\text{grad}f(p)`` is computed from ``∇f(p)`` as
+````math
+    \text{grad}f(p) = ∇f(p)p^Tp + Q_{2n}p∇f(p)^TQ_{2n}p,
+````
+where ``Q_{2n}`` is the [`SymplecticMatrix`](@ref).
+"""
+function gradient(M::SymplecticStiefel, f, p, backend::RiemannianProjectionBackend)
+    amb_grad = _gradient(f, p, backend.diff_backend)
+    return grad_euclidean_to_manifold(M, p, amb_grad)
+end
+
+function gradient!(M::SymplecticStiefel, f, X, p, backend::RiemannianProjectionBackend)
+    _gradient!(f, X, p, backend.diff_backend)
+    return grad_euclidean_to_manifold!(M, X, p, copy(X))
+end
+
+function grad_euclidean_to_manifold(M::SymplecticStiefel, p, ∇f_euc)
```

Comment:
We can do that, but how would one then 'manually' transform a known Euclidean gradient to the 
manifold gradient? I've tested and benchmarked a bit the difference between using `ForwardDiff` automatic differentiation and computing the Euclidean gradient "manually" and then transforming it using `grad_euclidean_to_manifold`, and there is quite a big difference when the matrices get bigger.

For example when projecting (200 x 40) matrices onto the symplectic manifold with manual conversion I get
a runtime on the order of 100 milli-seconds, whilst the runtime is closer to ten seconds on my machine when using the `Manifolds.ForwardDiffBackend` with the `gradient!(::SymplecticStiefel, f, p, X, ...)` API.

Reply:
gradient should do that – it is just calling this function – so I do not understand where your difference comes from? Ah I know. ForwardDiff does an approximation of the (Euclidean) Gradient with forwardDifference the wy you use it, so that is of course just an approximation to the actual gradient, we can discuss the details in our next meeting as well.